spec-tests/src/test/java/org/candlepin/spec/content/OwnerContentSpecTest.java

```diff
@@ -1157,4 +957,1008 @@ public void shouldPermitSafeParallelModification() throws Exception {
+                    .addProductContentItem(productContent);
+
+                contentMap.put(gprod.getId(), content);
+
+                subscriptions.add(this.createSubscription(globalOwner, gprod, false));
+                globalProducts.add(gprod);
+                this.mapProductInfo(gprod, null, null);
+            }
+
+            for (int oidx = 1; oidx <= owners.size(); ++oidx) {
+                OwnerDTO owner = owners.get(oidx - 1);
+
+                List<ProductDTO> ownerProducts = new ArrayList<>();
+                for (int i = 1; i <= 2; ++i) {
+                    ProductDTO cprod = new ProductDTO()
+                        .id(String.format("o%d-prod-%d", oidx, i))
+                        .name(String.format("%s product %d", owner.getKey(), i));
+
+                    cprod = this.adminClient.ownerProducts()
+                        .createProduct(owner.getKey(), cprod);
+
+                    ContentDTO content = this.adminClient.ownerContent()
+                        .createContent(owner.getKey(), Contents.random());
+
+                    this.adminClient.ownerProducts()
+                        .addContentToProduct(owner.getKey(), cprod.getId(), content.getId(), true);
+
+                    contentMap.put(cprod.getId(), content);
+
+                    ownerProducts.add(cprod);
+                    this.mapProductInfo(cprod, owner, null);
+                }
+
+                // Create an active and inactive global subscription for this org
+                subscriptions.add(this.createSubscription(owner, globalProducts.get(0), true));
+                subscriptions.add(this.createSubscription(owner, globalProducts.get(1), false));
+                this.mapProductInfo(globalProducts.get(0), null, owner);
+
+                // Create an active and inactive custom subscription
+                this.createPool(owner, ownerProducts.get(0), true);
+                this.createPool(owner, ownerProducts.get(1), false);
+                this.mapProductInfo(ownerProducts.get(0), owner, owner);
+            }
+
+            this.commitGlobalSubscriptions(subscriptions);
+        }
+
+        // Impl note:
+        // Since we cannot depend on the global state being prestine from run to run (or even test to
+        // test), we cannot use exact matching on our output as extraneous products from previous test
+        // runs or otherwise pre-existing data may show up in the result set. Instead, these tests will
+        // verify our expected products are present, and unexpected products from the test data are not.
+
+        private List<String> getUnexpectedIds(List<String> expectedPids) {
+            List<String> unexpectedPid = this.productMap.values()
+                .stream()
+                .map(ProductInfo::dto)
+                .map(ProductDTO::getId)
+                .filter(Predicate.not(expectedPids::contains))
+                .toList();
+
+            List<String> unexpectedCids = new ArrayList<>();
```

Comment:
Same here. You can either chain directly off the first stream (which should probably look closer to the previous example I provided to avoid the n*m lookup on contains), or you could call into the getContentIdsFromPids after resolving the unexpected products.

Any time you find yourself modifying or building a collection outside the scope of a functional block of calls, there's a chance something may not be quite right.

Reply:
Updated